include/wx/gtk/control.h

```diff
@@ -87,6 +87,13 @@ class WXDLLIMPEXP_CORE wxControl : public wxControlBase
     wxPoint GTKGetEntryMargins(GtkEntry* entry) const;
 
 private:
+#ifdef __WXGTK3__
+    // The cached result of GTKGetPreferredSize. The size by default is of
+    // wxSize(0, 0) and will be validated when GTKGetPreferredSize is called
+    // for the visible contol.
+    wxSize m_cachedPreferredSize;
```

Comment:
Thanks, I didn't think about this scenario. I guess we need to make `InvalidateBestSize()` virtual and override it to invalidate `m_cachedPreferredSize` too, even if this will just result in `(-1,-1)` being returned and still not anything useful -- but at least it won't be wrong.

I'd like to avoid having to make `InvalidateBestSize()` virtual because it feels wrong to allow overriding invalidating the cache managed entirely by the base class, but the only alternative I see is to make `Hide()` invalidate `m_cachedPreferredSize` if `m_bestSizeCache` is invalid. I think this should work (please correct me if I'm missing another case in which it won't?), but it looks a bit weird. Still arguably worth it though.

Reply:
FWIW, I agree, thanks for noticing this.

And, moreover, looking at this again, I think it should be called `m_preferredSizeCache` for consistency with `m_bestSizeCache` (even if the current name reads better). Could you please rename it when you make it mutable too?

Reply:
@vadz Ok, I'll rename it.
But with virtual `InvalidateBestSize` can be a problem if we will update `m_bestSizeCache` only from `InvalidateBestSize`. Example:
```c++
control->GetBestSize();
... // do something to change best size
control->Hide();
control->InvalidateBestSize(); // Don't update the cache because the control is hidden
control->Show();
control->GetBestSize();  // GTKGetPreferredSize() returns outdated size
```

Reply:
Invalidating `m_preferredSizeCache` in `InvalidateBestSize()` would make it the same thing as `m_bestSizeCache`. I don't think the original problem would be solved.

Reply:
Yes, you're both right, sorry. It looks like we actually need to call `InvalidateBestSize()` from `Show()` in wxGTK if the best size is the cached preferred size. I.e. we need another flag just to remember whether this is the case, i.e. set it if we ever actually use `m_preferredSizeCache` (as this means that `DoGetBestSize()` was called while the window was hidden).

But speaking of the original problem, I've actually realized that I'm not so sure about it any more: if we call `InvalidateBestSize()` on the editor window in wxGrid, it means that its best size must have changed (e.g. because its contents changed, or because its font changed or whatever). And if the editor window is hidden, we simply have no way of computing the actual best size. So how does caching the preferred size actually help us at all? Is it just better than nothing in the sense that it's still wrong but less wrong than just returning invalid or null size?

Reply:
> Is it just better than nothing in the sense that it's still wrong but less wrong than just returning invalid or null size?

Looks like it is.